arch/cortex-m/samv71-hal/src/uart/reader.rs

```diff
@@ -0,0 +1,114 @@
+/// Once taken, it can be put inside UART again using [`UART::put_reader`](super::UART::put_reader)
+/// for storage.
+///
+/// # Safety
+/// If Reader is used while UART receiver is disabled, it will always return [`Error::TimedOut`] on
+/// blocking operations.
+///
+/// Reader is thread-safe, as it doesn't share any (mutable) state with UART or Writer, and
+/// there can be only a single instance of Reader per UART.
+pub struct Reader<Instance: UARTMetadata> {
+    /// UART instance marker.
+    _uart: PhantomData<Instance>,
+}
+
+impl<Instance: UARTMetadata> Reader<Instance> {
+    /// Receives a single byte. Blocks until a byte is received, or timeout is hit.
+    ///
+    /// If you check "receiver ready" flag manually (for example, in IRQ handler), you could use
+    /// [`Reader::get_received_byte`] instead, as it doesn't perform the additional status check.
+    /// However, this function will also work fine in that context, it'll just double-check that.
+    ///
+    /// This function requires mutable access to Reader, as reading the character from RX holding
+    /// register while "receiver ready" flag is set will reset it's state and clear this flag.
+    ///
+    /// # Parameters
+    /// * `timeout` - Maximum amount of UART status checks before declaring timeout.
+    ///
+    /// # Returns
+    /// `Ok(u8)` if reception was successful, with the value of received byte.
+    /// `Err(())` on timeout.
+    pub fn receive_byte(&mut self, timeout: u32) -> Result<u8, Error> {
+        self.wait_for_byte_reception(timeout)
+            // This is safe, as we just verified that receiver is ready and RX holding register
+            // contains a received byte.
+            .map_or(Err(Error::TimedOut), |_| unsafe {
+                Ok(self.get_received_byte())
+            })
+    }
+
+    /// Returns the byte currently stored in received character register.
+    ///
+    /// This function is meant to be used primarily in interrupt handlers, as a slightly faster
+    /// version of [`Reader::receive_byte`] that avoids double-checking the status register.
+    ///
+    /// This function requires mutable access to Reader, as reading the character from RX holding
+    /// register while "receiver ready" flag is set will reset it's state and clear this flag.
+    ///
+    /// # Safety
+    /// This function doesn't wait for UART to indicate that there's data in RX register, and will
+    /// return `0` if there's no received data there, instead of an error.
+    /// Therefore, it's reasonable to use only if you manually checked if there's new data in UART
+    /// RX register (by checking "receiver ready" status flag). If you do that, then this function
+    /// becomes safe to use.
+    ///
+    /// # Returns
+    /// Received byte, if UART status flag indicates that there's one in RX register.
+    /// `0`` otherwise.
+    #[inline(always)]
+    pub unsafe fn get_received_byte(&mut self) -> u8 {
```

Comment:
That's the reason for `unsafe` right there.
To make this function safe, it would have to read status register. `receive_byte` does that. This is just a simple abstraction over the register operation that should be used in interrupts, and you'd usually read status manually in them anyway, so `receive_byte` would just add another unnecessary read to that.